### PR TITLE
Bybit createOrder (stopLossPrice, takeProfitPrice)

### PR DESCRIPTION
**Pull request by Dan-krm**
Added stopLoss and takeProfit functionality to Bybit contract and USDC swap markets:
#13822

createContractOrder:
```
node examples/js/cli bybit createOrder BTC/USDT:USDT limit buy 0.002 20000 '{"takeProfitPrice":"25000","stopLossPrice":"19000"}'

bybit.createOrder (BTC/USDT:USDT, limit, buy, 0.002, 20000, [object Object])
2022-06-15T07:18:30.335Z iteration 0 passed in 304 ms

{
  info: {
    order_id: '6998b879-b247-47bf-a829-5a6dae436d59',
    user_id: '3172070',
    symbol: 'BTCUSDT',
    side: 'Buy',
    order_type: 'Limit',
    price: '20000',
    qty: '0.002',
    time_in_force: 'GoodTillCancel',
    order_status: 'Created',
    last_exec_price: '0',
    cum_exec_qty: '0',
    cum_exec_value: '0',
    cum_exec_fee: '0',
    reduce_only: false,
    close_on_trigger: false,
    order_link_id: '',
    created_time: '2022-06-15T07:18:30Z',
    updated_time: '2022-06-15T07:18:30Z',
    take_profit: '25000',
    stop_loss: '19000',
    tp_trigger_by: 'LastPrice',
    sl_trigger_by: 'LastPrice',
    position_idx: '1'
  },
  id: '6998b879-b247-47bf-a829-5a6dae436d59',
  clientOrderId: undefined,
  timestamp: 1655277510000,
  datetime: '2022-06-15T07:18:30.000Z',
  lastTradeTimestamp: 1655277510000,
  symbol: 'BTC/USDT:USDT',
  type: 'limit',
  timeInForce: 'GTC',
  postOnly: false,
  side: 'buy',
  price: 20000,
  stopPrice: undefined,
  stopLossPrice: '19000.00',
  takeProfitPrice: '25000.00',
  amount: 0.002,
  cost: 0,
  average: undefined,
  filled: 0,
  remaining: 0.002,
  status: 'open',
  fee: { cost: 0, currency: 'USDT' },
  trades: [],
  fees: [ { cost: 0, currency: 'USDT' } ]
}
```

createUsdcOrder:
```
node examples/js/cli bybit createOrder BTCPERP limit buy 0.011 20000 '{"takeProfitPrice":"25000","stopLossPrice":"19000"}'

bybit.createOrder (BTCPERP, limit, buy, 0.011, 20000, [object Object])
2022-06-15T07:56:43.463Z iteration 0 passed in 309 ms

{
  info: {
    orderId: '923f538b-224c-4ea3-988e-351c96f3a98a',
    orderLinkId: '',
    mmp: false,
    symbol: 'BTCPERP',
    orderType: 'Limit',
    side: 'Buy',
    orderQty: '0.01100000',
    orderPrice: '20000.00',
    iv: '0',
    timeInForce: 'GoodTillCancel',
    orderStatus: 'Created',
    createdAt: '1655279803742258',
    basePrice: '0.00',
    triggerPrice: '0.00',
    takeProfit: '25000.00',
    stopLoss: '19000.00',
    slTriggerBy: 'MarkPrice',
    tpTriggerBy: 'MarkPrice'
  },
  id: '923f538b-224c-4ea3-988e-351c96f3a98a',
  clientOrderId: undefined,
  timestamp: 1655279803742,
  datetime: '2022-06-15T07:56:43.742Z',
  lastTradeTimestamp: undefined,
  symbol: 'BTC/USD:USDC',
  type: 'limit',
  timeInForce: 'GTC',
  postOnly: false,
  side: 'buy',
  price: 20000,
  stopPrice: '0.00',
  stopLossPrice: '19000.00',
  takeProfitPrice: '25000.00',
  amount: 0.011,
  cost: undefined,
  average: undefined,
  filled: undefined,
  remaining: undefined,
  status: 'open',
  fee: undefined,
  trades: [],
  fees: []
}
```

### CODE REVIEW

File: js/bybit.js [hunk 1]
```diff
@@ -2642,6 +2642,8 @@ module.exports = class bybit extends Exchange {
             'side': side,
             'price': price,
             'stopPrice': stopPrice,
+            'stopLossPrice': this.safeNumber2 (order, 'stop_loss', 'stopLoss'),
```
Comment (carlosmiei) [inline js/bybit.js:2645]: When not set they default to 0 so I guess we should make it undefined in this case

```suggestion
            let stopLossPrice = this.safeNumber2 (order, 'stop_loss', 'stopLoss');
            if (stopLossPrice === 0) {
                stopLossPrice = undefined;
            }
```

File: js/bybit.js [hunk 2]
```diff
@@ -2933,19 +2954,37 @@ module.exports = class bybit extends Exchange {
         if (price !== undefined) {
             request['price'] = parseFloat (this.priceToPrecision (symbol, price));
         }
-        const stopPx = this.safeValue2 (params, 'stop_px', 'stopPrice');
+        const triggerPrice = this.safeValue2 (params, 'stop_px', 'stopPrice');
```
Comment (carlosmiei) [inline js/bybit.js:2957]: Shouldn't we accept 'triggerPrice' here as well?

File: js/bybit.js [hunk 3]
```diff
@@ -2933,19 +2954,37 @@ module.exports = class bybit extends Exchange {
...
+                if (basePrice === undefined) {
+                    throw new ArgumentsRequired (this.id + ' createOrder() requires both the stop_px and base_price params for a conditional ' + type + ' order');
+                }
+                const triggerBy = this.safeString2 (params, 'trigger_by', 'triggerBy', 'LastPrice');
+                request['trigger_by'] = triggerBy;
+                request['stop_px'] = parseFloat (this.priceToPrecision (symbol, triggerPrice));
+                request['base_price'] = parseFloat (this.priceToPrecision (symbol, basePrice));
+            }
+            if (isStopLossOrder) {
+                request['stop_loss'] = stopLossPrice;
```
Comment (carlosmiei) [inline js/bybit.js:2978]: missing **priceToPrecision** here

File: js/bybit.js [hunk 4]
```diff
@@ -2933,19 +2954,37 @@ module.exports = class bybit extends Exchange {
...
+                const slTriggerBy = this.safeString2 (params, 'sl_trigger_by', 'LastPrice');
+                request['sl_trigger_by'] = slTriggerBy;
+            }
+            if (isTakeProfitOrder) {
+                request['take_profit'] = takeProfitPrice;
```
Comment (carlosmiei) [inline js/bybit.js:2983]: here too 🙂

File: js/bybit.js (hunk 2 above)
Comment (Dan-krm) [reply js/bybit.js:2965]: I'm not sure, but we unified stopPrice and sometimes the exchange specific param for stopPrice is triggerPrice, in this case its stop_px.

If I were to add 'triggerPrice' here, what would be the best approach? can I have three params in safeValue2? should it be safeValue with the argument being a list of all three potential params, or is there something else I can use in this situation?

**Review by carlosmiei — APPROVED** [review]
Everything seems good now ✅

**Comment by frosty00** [comment]
have made edits, will merge in an hour

**Comment by kroitor** [comment]
We need to automate order testing.

**Comment by frosty00** [comment]
> We need to automate order testing.

yeah, but it's hard :), especially to automate the stopLoss price checking

**Comment by frosty00** [comment]
anyway the reason for my edits:

previously dan's PR made a limit order with an attached tp/sl order which is trigger after the limit order is opened. however, there is also the option on bybit to make a regular trigger order which triggers when the prices crosses a certain amount. this is the unified stopLoss and takeProfit, not the first one.

**Comment by frosty00** [comment]
in this case i will merge, since it fixes some things, including allowing users to create usdc orders (was broken cause it was missing the required param 'OrderFilter' in master)

**Comment by kroitor** [comment]
> > We need to automate order testing.
> 
> yeah, but it's hard :), especially to automate the stopLoss price checking

It should not crash upon placing such orders and any edit to the stop loss and take profit functionality should not affect regular orders and vice versa – having at least that kind of testing would be a huge progress.

**Comment by kroitor** [comment]
> in this case i will merge, since it fixes some things, including allowing users to create usdc orders (was broken cause it was missing the required param 'OrderFilter' in master)

Please, check discord first.

File: js/bybit.js [hunk 5]
```diff
@@ -2925,41 +2951,57 @@ module.exports = class bybit extends Exchange {
...
         }
         if (market['linear']) {
-            const reduceOnly = this.safeValue2 (params, 'reduce_only', 'reduceOnly', false);
-            const closeOnTrigger = this.safeValue2 (params, 'reduce_only', 'reduceOnly', false);
-            request['reduce_only'] = reduceOnly;
-            request['close_on_trigger'] = closeOnTrigger;
-        }
-        if (price !== undefined) {
-            request['price'] = parseFloat (this.priceToPrecision (symbol, price));
+            request['reduce_only'] = this.safeValue2 (params, 'reduce_only', 'reduceOnly', false);
```
Comment (samgermain) [inline js/bybit.js:2954]: this sends `"reduce_only": undefined` to the endpoint if the user doesn't pass it in. I thought that was something undesirable that we didn't want to do

@kroitor your opinion?

Comment (carlosmiei) [reply js/bybit.js:2954]: @samgermain 'false' is passed as the default value, no?

**Comment by Senix4** [comment]
does this also work with market orders? so basically opening a position by market order and closing it by setting take profit and stop loss (limit orders)?

**Comment by Dan-krm** [comment]
> does this also work with market orders? so basically opening a position by market order and closing it by setting take profit and stop loss (limit orders)?

@Senix4 Yeah the stopLossPrice and takeProfitPrice are used after opening a position so they can be used after opening a position with a market order, you can read more about the different trigger, stop loss and take profit order types that are supported here: https://github.com/ccxt/ccxt/wiki/Manual#placing-orders

**Pull request merged**